File: frecuencyCounterPattern/same/same.rb
def same(arr1, arr2)
    return false if arr1.length != arr2.length

    frecuency_counter1 = {}
    frecuency_counter2 = {}

    arr1.each { |elm| frecuency_counter1[elm] = (frecuency_counter1[elm] || 0) + 1 }
    arr2.each { |elm| frecuency_counter2[elm] = (frecuency_counter2[elm] || 0) + 1 }

    frecuency_counter1.each  do |elm, counter|
        return false if !frecuency_counter2.include?(elm ** 2)
        return false if frecuency_counter2[elm ** 2] != counter
    end

    return true
end

puts(same([1, 2, 3, 4], [1, 3, 9, 1])) # false

File: frecuencyCounterPattern/same/same.py
def same(arr1, arr2):
    if len(arr1) != len(arr2):
        return False

    frequecy_counter1 = {}
    frequecy_counter2 = {}

    for elm in arr1:
        frequecy_counter1[elm] = frequecy_counter1.get(elm, 0) + 1

    for elm in arr1:
        frequecy_counter2[elm ** 2] = frequecy_counter2.get(elm ** 2, 0) + 1

    for elm, counter in frequecy_counter1.items():
        if elm ** 2 not in frequecy_counter2:
            return False

        if frequecy_counter2.get(elm ** 2) != counter:
            return False

    return True


print(same([1, 2, 3, 4], [1, 4, 9, 16])) # True
